### PR TITLE
BUG: fix `ma.average` not working well with `nan` weights

### LINKED ISSUE #20375: DOC: The np.ma.average documentation does not account for cases where weights contain np.nan.
### Issue with current documentation:

If the input weights array contains `np.nan`, `ma.average` returns `np.nan` even if `np.nan` is in a masked position in the input array.
In the example code below, the `np.nan` in weights is in a masked position in the input array, and it returns `np.nan`

```python
np.ma.average(
    np.ma.array([1., 2., 3., 4.], mask=[False, False, True, True]),
    weights=np.array([1, 1, 1, np.nan]),
    returned=True
)
```

It seems that the `np.nan` in the masked position doesn't need to be summed in this [code](https://github.com/numpy/numpy/blob/v1.21.0/numpy/ma/extras.py#L623).
However, in this [line](https://github.com/numpy/numpy/blob/v1.21.0/numpy/ma/extras.py#L620-L621), `np.nan` is passed regardless of the mask.

Is this situation intentional? Or is it something that needs to be fixed?
If intended, how about adding an example or additional explanation for this situation?

### Idea or request for content:

_No response_

### PR DESCRIPTION
**Pull request by scratchmex**
If a weight is nan it behaves erratically. See the tests added.

closes #20375

### CODE REVIEW

**Review by melissawm — APPROVED** [review]
This looks good to me, thanks @scratchmex - will give other maintainers a chance to look before merging.

File: numpy/ma/extras.py
Comment (mattip) [inline numpy/ma/extras.py:595]: I think this could have side effects for things like cupy arrays. Is it required?

Comment (scratchmex) [reply numpy/ma/extras.py:595]: Yes because later I add a mask to the weights array.

Comment (mattip) [reply numpy/ma/extras.py:595]: I see there are many more instances of `asarray` than of `asanyarray` across the masked array implementation. So this problem is not unique to this PR.

**Comment by mattip** [comment]
Thanks @scratchmex

**Comment by scratchmex** [comment]
@charris @mattip just wondering if this is a backport candidate

**Comment by Junyeopp** [comment]
@mattip @scratchmex Is there any updates?

**Comment by mattip** [comment]
@Junyeopp this will be part of the upcoming 1.23 release. You can already use the release candidates via `python -m pip install numpy==1.23.0rc3`

**Pull request merged**
